text/0000-int-guidelines.md

```diff
@@ -0,0 +1,164 @@
+- Start Date: 2015-01-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes usage guidelines for the various integer types.
+
+# Motivation
+
+The goal of this RFC is to help people decide what integer types to use when they need
+to make a decision for a new API.
+
+It builds on [https://github.com/rust-lang/rfcs/pull/560](the integer overflow RFC),
+which provides debug-time assertions for overflow and underflow. One of the goals
+of this RFC is to provide guidance for when to use unsigned types, and these
+assertions affect the traditional tradeoffs about unsigned types, which are described
+under the detailed design below.
+
+It also draws inspiration from the [Google C++ Style Guide](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types)
+
+This RFC attempts to balance several concerns:
+
+* Rust developers should have an easy, go-to heuristic for deciding what
+  integer sizes to use when they need to use a number type.
+* 32-bit integers are considerably faster than 64-bit integers in some
+  situations, so using 64-bit integers for tiny numbers, especially in
+  hot code, can result in unnecessarily slower programs.
+* Reflexive use of 32-bit integers too-often results in overflows when the
+  numbers aren't expected to be "laughably smaller" than the maximum
+  32-bit number.
+* Occasional use of `usize` when building a brand new data structure may
+  be appropriate, but use of `usize` in general can introduce portability
+  hazards when the use-case is not proportional to the amount of
+  addressable memory.
+* Using of unsigned integers is traditionally thought to be error-prone, and
```

Comment:
More aspects about unsigned integers:

* Unsigned integers are fitting for bit patterns, hash codes, wraparound mod 2^N, and values that aren't really numbers.
* Since Rust requires explicit conversions from signed to unsigned integers, some C/C++ cautionary cases don't apply. [My mistake in not realizing this sooner.] E.g. it's tempting to make a C function take an uint32_t so it doesn't have to deal with negative inputs but it's easy in C to accidentally pass in a negative integer. Now the function can't detect that bad input unless it has a suitable upper bound.  
[BTW, "polymorphic array indexing" would make it OK to index a Rust array with a signed integer. Rustc could implement a signed bounds check with a single unsigned comparison (at `usize` or larger width) since the upper bound cannot exceed `isize::MAX`.]
* Hopefully rustc will always complain about `i >= 0` for unsigned `i`.
* There are plenty of cases to be careful about exceeding the unsigned conceptual domain. Making a value unsigned looks better than it works out.